File: 11-stacks-queues2/3-monotonic-queues/main.py
# Monotonic Queue - Python Example

from collections import deque

def max_sliding_window(nums, k):
    dq = deque()
    result = []

    for i, num in enumerate(nums):
        while dq and nums[dq[-1]] < num:
            dq.pop()
        dq.append(i)

        if dq[0] == i - k:
            dq.popleft()

        if i >= k - 1:
            result.append(nums[dq])

    return result

nums = [1,3,-1,-3,5,3,6,7]
k = 3
print("Sliding window maximum:", max_sliding_window(nums, k))

File: 11-stacks-queues2/3-monotonic-queues/main.java
// Monotonic Queue - Java Example

import java.util.Deque;
import java.util.LinkedList;
import java.util.ArrayList;
import java.util.List;

public class MonotonicQueue {

    public static List<Integer> maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> dq = new LinkedList<>();
        List<Integer> result = new ArrayList<>();

        for (int i = 0; i < nums.length; i++) {
            while (!dq.isEmpty() && nums[dq.peekLast()] < nums[i]) {
                dq.pollLast();
            }
            dq.offerLast(i);

            if (dq.peekFirst() == i - k) {
                dq.pollFirst();
            }

            if (i >= k - 1) {
                result.add(nums[dq.peekFirst()]);
            }
        }
        return result;
    }

    public static void main(String[] args) {
        int[] nums = {1,3,-1,-3,5,3,6,7};
        int k = 3;
        List<Integer> maxVals = maxSlidingWindow(nums, k);
        System.out.print("Sliding window maximum: ");
        for (int val : maxVals) {
            System.out.print(val + " ");
        }
        System.out.println();
    }
}
